builder/parser/line_parsers.go

```diff
@@ -195,6 +192,44 @@ func parseLabel(rest string) (*Node, map[string]bool, error) {
 	return parseNameVal(rest, "LABEL")
 }
 
+// parses a statement with multiple keywords definition and name-value
+// assignment, like `name1= name2=value`.
+// Note that this is a stricter format than the old format of assignment, in a way
+// that it only allows assignment of the form `name=value`. And that a name
+// definition or assignment to an empty value is designated by `name=`.
+func parseNameOrNameVal(rest string) (*Node, map[string]bool, error) {
+	words := parseWords(rest)
+	if len(words) == 0 {
+		return nil, nil, nil
+	}
+
+	var (
+		rootnode *Node
+		prevNode *Node
+	)
+	for i, word := range words {
+		if !strings.Contains(word, "=") {
+			return nil, nil, fmt.Errorf("Syntax error - can't find = in %q. Must be of the form: name=value", word)
```

Comment:
@duglin 
> While playing with this I think printing the error here might not be ideal.
> ...
> By not seeing the steps it would make it even harder to know which cmd caused this error.

I followed a similar error handling pattern as it exists in `parseNameVal` i.e. it returns an error if the passed line doesn't follow the prescribed name-val format.

But I do see your concern with this error being confusing as it doesn't give much context on the failed Dockerfile statement. To make this error more useful I can pass the `ARG` string to the `parseNameOrNameVal` function (similar to what `parseEnv` and `parseLabel` do) and add it in the error string. This will work as well, right?

> Would it make more sense to have the parser accept "foo", "foo=" and "foo=bar" but then have the arg() func, in dispatcher.go, check for a "=" and generate a more focused error message?
> Additionally, I'd prefer if the parser wasn't this smart about what it was parsing. We should be moving away from a smart parser (that knows about Dockerfile commands) and one that just knows about words/spaces/etc.

I agree, about not having a lot of intelligence related to specific Dockerfile primitive in the parsers. However, the intelligence that I foresee in `parseNameOrNameVal` is partly what you described i.e. it should accept `foo`, `foo=` and `foo=bar` and in addition normalize the names to a consistent string i.e. both `foo` and `foo=` shall return the name `foo` after parsing. Accordingly, my only concern with checking `=` in `arg()` instead is that it will move parsing logic inside the dispatcher(s) and I feel this is a parsing error than a dispatch error.